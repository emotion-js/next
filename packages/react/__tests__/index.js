import React from 'react'
import serializer from 'jest-glamor-react'
import { css, sheet } from 'new-css-in-js'
import styled from 'react-new-css-in-js'
import { render, mount } from 'enzyme'

expect.addSnapshotSerializer(serializer(sheet))

describe('styled', () => {
  test('no dynamic', () => {
    const H1 = styled('h1')`float: left;`

    const tree = render(<H1>hello world</H1>)

    expect(tree).toMatchSnapshot()
  })

  test('basic render', () => {
    const fontSize = 20
    const H1 = styled('h1')`
      color: blue;
      font-size: ${fontSize};
      @media (min-width: 420px) {
        color: blue;
        @media (min-width: 520px) {
          color: green;
        }
      }
    `

    const tree = render(<H1>hello world</H1>)

    expect(tree).toMatchSnapshot()
  })

  test('nested', () => {
    const fontSize = '20px'
    const H1 = styled('h1')`font-size: ${fontSize};`

    const Thing = styled('div')`
      display: flex;
      & div {
        color: green;

        & span {
          color: red;
        }
      }
    `

    const tree = render(
      <Thing>
        hello <H1>This will be green</H1> world
      </Thing>
    )

    expect(tree).toMatchSnapshot()
  })

  test.skip('random expressions', () => {
    const margin = (t, r, b, l) => {
      return props => css`
        margin-top: ${t};
        margin-right: ${r};
        margin-bottom: ${b};
        margin-left: ${l};
      `
    }

    const mq = css`
      @media (min-width: 420px) {
        color: blue;
        @media (min-width: 520px) {
          color: green;
        }
      }
    `

    const H1 = styled('h1')`
      ${mq};
      ${props => props.prop && css`font-size: 1rem;`};
      ${margin(0, 'auto', 0, 'auto')};
      color: green;
    `

    const tree = render(
      <H1 className={'legacy__class'} prop>
        hello world
      </H1>
    )

    expect(tree).toMatchSnapshot()
  })

  test('random expressions undefined return', () => {
    const H1 = styled('h1')`
      ${props => props.prop && css`font-size: 1rem;`};
      color: green;
    `

    const tree = render(<H1 className={'legacy__class'}>hello world</H1>)

    expect(tree).toMatchSnapshot()
  })

  test('random object expression', () => {
    const margin = (t, r, b, l) => {
      return props => ({
        marginTop: t,
        marginRight: r,
        marginBottom: b,
        marginLeft: l
      })
    }
    const H1 = styled('h1')`
      background-color: hotpink;
      ${props => props.prop && { fontSize: '1rem' }};
      ${margin(0, 'auto', 0, 'auto')};
      color: green;
    `

    const tree = render(
      <H1 className={'legacy__class'} prop>
        hello world
      </H1>
    )

    expect(tree).toMatchSnapshot()
  })

  test('composition', () => {
    const fontSize = 20
    const H1 = styled('h1')`font-size: ${fontSize + 'px'};`

    const H2 = styled(H1)`font-size: ${fontSize * 2 / 3};`

    const tree = render(<H2 className={'legacy__class'}>hello world</H2>)

    expect(tree).toMatchSnapshot()
  })

  test('selectivity when composing', () => {
    const red = css`color: red;`
    const blue = css`color: blue;`
    const green = css`color: green;`

    const final = css`
      ${green};
      ${blue};
      ${red};
    `

    const H1 = styled('h1')`${final};`
    const H2 = styled(H1)`color: purple;`

    const tree = render(<H2 className={green}>What color am I?</H2>)

    expect(tree).toMatchSnapshot()
  })

  test('input placeholder', () => {
    const Input = styled('input')`
      ::placeholder {
        background-color: green;
      }
    `
    const tree = render(<Input />)

    expect(tree).toMatchSnapshot()
  })

  test('handles more than 10 dynamic properties', () => {
    const H1 = styled('h1')`
      text-decoration: ${'underline'};
      border-right: solid blue 54px;
      background: ${'white'};
      color: ${'black'};
      display: ${'block'};
      border-radius: ${'3px'};
      padding: ${'25px'};
      width: ${'500px'};
      z-index: ${100};
      font-size: ${'18px'};
      text-align: ${'center'};
      border-left: ${p => p.theme.blue};
    `

    const tree = render(
      <H1 className={'legacy__class'} theme={{ blue: 'blue' }}>
        hello world
      </H1>
    )

    expect(tree).toMatchSnapshot()
  })

  test('function in expression', () => {
    const fontSize = 20
    const H1 = styled('h1')`font-size: ${fontSize + 'px'};`

    const H2 = styled(H1)`font-size: ${({ scale }) => fontSize * scale + 'px'};`

    const tree = render(
      <H2 scale={2} className={'legacy__class'}>
        hello world
      </H2>
    )

    expect(tree).toMatchSnapshot()
  })

  test('innerRef', () => {
    const H1 = styled('h1')`font-size: 12px;`

    const refFunction = jest.fn()

    const tree = mount(<H1 innerRef={refFunction}>hello world</H1>)

    expect(tree).toMatchSnapshot()
    expect(refFunction).toBeCalled()
  })

  test('composing components', () => {
    const Button = styled('button')`color: green;`
    const OtherButton = styled(Button)`display: none;`

    const AnotherButton = styled(OtherButton)`
      display: flex;
      justify-content: center;
    `
    const tree = render(<AnotherButton>hello world</AnotherButton>)

    expect(tree).toMatchSnapshot()
  })

  test('prop filtering', () => {
    const Link = styled('a')`color: green;`
    const rest = { m: [3], pt: [4] }

    const tree = render(
      <Link
        a
        b
        wow
        prop
        filtering
        is
        cool
        aria-label="some label"
        data-wow="value"
        href="link"
        {...rest}
      >
        hello world
      </Link>
    )

    expect(tree).toMatchSnapshot()
  })
  test('no prop filtering on non string tags', () => {
    const Link = styled(props => <a {...props} />)`color: green;`

    const tree = render(
      <Link
        a
        b
        wow
        prop
        filtering
        is
        cool
        aria-label="some label"
        data-wow="value"
        href="link"
      >
        hello world
      </Link>
    )

    expect(tree).toMatchSnapshot()
  })
  test('prop filtering on composed styled components that are string tags', () => {
    const BaseLink = styled('a')`background-color: hotpink;`
    const Link = styled(BaseLink)`color: green;`

    const tree = render(
      <Link
        a
        b
        wow
        prop
        filtering
        is
        cool
        aria-label="some label"
        data-wow="value"
        href="link"
      >
        hello world
      </Link>
    )

    expect(tree).toMatchSnapshot()
  })
})
