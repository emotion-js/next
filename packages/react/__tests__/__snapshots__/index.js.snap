// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled basic render 1`] = `
.glamor-0 {
  color: blue;
  font-size: ;
}

@media (min-width:520px) {
  .glamor-0 {
    color: green;
  }
}

@media (min-width:420px) {
  .glamor-0 {
    color: blue;
  }
}

<h1
  class="glamor-0"
>
  hello world
</h1>
`;

exports[`styled composing components 1`] = `
.glamor-0 {
  color: green;
  display: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<button
  class="glamor-0"
>
  hello world
</button>
`;

exports[`styled composition 1`] = `
.glamor-0 {
  font-size: 20px;
  font-size: ;
}

<h1
  class="glamor-0 legacy__class"
>
  hello world
</h1>
`;

exports[`styled function in expression 1`] = `
.glamor-0 {
  font-size: 20px;
  font-size: 40px;
}

<h1
  class="glamor-0 legacy__class"
>
  hello world
</h1>
`;

exports[`styled handles more than 10 dynamic properties 1`] = `
.glamor-0 {
  text-decoration: underline;
  border-right: solid blue 54px;
  background: white;
  color: black;
  display: block;
  border-radius: 3px;
  padding: 25px;
  width: 500px;
  z-index: ;
  font-size: 18px;
  text-align: center;
  border-left: blue;
}

<h1
  class="glamor-0 legacy__class"
>
  hello world
</h1>
`;

exports[`styled innerRef 1`] = `
.glamor-0 {
  font-size: 12px;
}

<Styled
  innerRef={[Function]}
>
  <h1
    className="glamor-0"
  >
    hello world
  </h1>
</Styled>
`;

exports[`styled input placeholder 1`] = `
.glamor-0::placeholder {
  background-color: green;
}

<input
  class="glamor-0"
/>
`;

exports[`styled nested 1`] = `
.glamor-1 div span {
  color: red;
}

.glamor-1 div {
  color: green;
}

.glamor-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.glamor-0 {
  font-size: 20px;
}

<div
  class="glamor-1"
>
  hello
  <h1
    class="glamor-0"
  >
    This will be green
  </h1>
   world
</div>
`;

exports[`styled no dynamic 1`] = `
.glamor-0 {
  float: left;
}

<h1
  class="glamor-0"
>
  hello world
</h1>
`;

exports[`styled no prop filtering on non string tags 1`] = `
<a
  a="true"
  aria-label="some label"
  b="true"
  classname="css-1fpqmhj"
  cool="true"
  data-wow="value"
  filtering="true"
  href="link"
  is="true"
  prop="true"
  wow="true"
>
  hello world
</a>
`;

exports[`styled prop filtering 1`] = `
<a
  aria-label="some label"
  classname="css-1fpqmhj"
  data-wow="value"
  href="link"
  is="true"
>
  hello world
</a>
`;

exports[`styled prop filtering on composed styled components that are string tags 1`] = `
<a
  aria-label="some label"
  classname="css-1tp52ze"
  data-wow="value"
  href="link"
  is="true"
>
  hello world
</a>
`;

exports[`styled random expressions undefined return 1`] = `
.glamor-0 {
  color: green;
}

<h1
  class="glamor-0 legacy__class"
>
  hello world
</h1>
`;

exports[`styled random object expression 1`] = `
.glamor-0 {
  background-color: hotpink;
  font-size: 1rem;
  margin-right: auto;
  margin-left: auto;
  color: green;
}

<h1
  class="glamor-0 legacy__class"
>
  hello world
</h1>
`;
